### PR TITLE
test: add Between UT for type_coercion

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Related #5928 

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

add UT.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by jackwener** [comment]
We can do some improvement for Type coercion. 
I have some draft for them. I need some time to think and polish them

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 1]
```diff
@@ -999,8 +1000,56 @@ mod test {
...
+                    data_type: DataType::Date32,
+                })),
+                op: Operator::Plus,
+                right: Box::new(Expr::Literal(ScalarValue::IntervalYearMonth(Some(1)))),
+            })),
+        ));
+        let empty = empty_with_type(DataType::Utf8);
+        let plan = LogicalPlan::Filter(Filter::try_new(expr, empty)?);
+        let expected =
+            "Filter: a BETWEEN Utf8(\"2002-05-08\") AND CAST(CAST(Utf8(\"2002-05-08\") AS Date32) + IntervalYearMonth(\"1\") AS Utf8)\
```
Comment (yjshen) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:1025]: Dumb Question: while I was reading temporal coercion rules in https://github.com/apache/arrow-datafusion/blob/main/datafusion/expr/src/type_coercion/binary.rs#L688, I found UTF8 vs Date comparison always yields Date type. Why are we expecting upper bounds to be cast as UTF8?

Comment (jackwener) [reply datafusion/optimizer/src/analyzer/type_coercion.rs:1025]: I think It isn't valid. Current many problem exist in `TypeCoercion`, I'm fixing them one by one.
BTW, I think I need to research and think about it systematically, not just patching case by case.

Comment (yjshen) [reply datafusion/optimizer/src/analyzer/type_coercion.rs:1025]: I suppose we could put the desired coercion as expected. But ignore this test case for now and link the issue trying to fix it?

BTW, I checked the temporal coercion rule in Spark and found its default coerce to datetime types after its 3.0 version, but left a legacy conf to make string vs date to coerce to string type. I agree with you that we should always compare using date types.

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 2]
```diff
@@ -999,8 +1000,56 @@ mod test {
...
             "Projection: CAST(a AS Decimal128(24, 4)) IN ([CAST(Int32(1) AS Decimal128(24, 4)), CAST(Int8(4) AS Decimal128(24, 4)), CAST(Int64(8) AS Decimal128(24, 4))]) AS a IN (Map { iter: Iter([Int32(1), Int8(4), Int64(8)]) })\
              \n  EmptyRelation";
-        assert_analyzed_plan_eq(Arc::new(TypeCoercion::new()), &plan, expected)?;
-        Ok(())
+        assert_analyzed_plan_eq(Arc::new(TypeCoercion::new()), &plan, expected)
+    }
+
+    #[test]
+    fn between_case() -> Result<()> {
+        let expr = Expr::Between(Between::new(
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:1008]: It is unfortunate it takes this much code to create a `between` -- I will make a follow on PR to reduce the boiler plate needed

**Pull request merged**